### ISSUE TITLE
Unit testing support

### ISSUE THREAD

**Issue by jeamland** [jemalloc/jemalloc#969]
So in a lot of C you get the common pattern of:

```
thing = malloc(intended_size_of_thing);
if (thing == NULL) {
    deal_with_lack_of_thing();
}
yay_i_have_a(thing);
```

Unit testing the failure path can be tricky. There exist debug mallocs that may help with this. I've also used `setrlimit()` to cause failures.

It would be a worthy addition to jemalloc's debugging support to have a configurable that caused the next call to `malloc()` (or any of the other functions) to fail.

**Comment by edsiper** [comment]
:+1:

**Comment by davidtgoldblatt** [comment]
It doesn't seem like there's that much of a reason for this to live inside jemalloc, as opposed to a separate debug library; it doesn't need to access internal data structures or the like. Is there something you had in mind that needs support on our end?

**Comment by jeamland** [comment]
So here's what I had in mind:

https://gist.github.com/jeamland/80dbd1bffae1612ed6972001ec4c6b71

This code does the right things when linked against a jemalloc containing the following commit:

https://github.com/jeamland/jemalloc/commit/6360738633baca49cab1e380a29bef7fd1c380f2

That's obviously a work in progress and all the names and whatnot are definitely up for discussion but it shows what I'd like to have.

**Comment by davidtgoldblatt** [comment]
Yeah, this definitely *could* be done within jemalloc. What I mean more is, how does life get better for someone by moving the fault-injection stuff out of a debug allocator and into jemalloc?

For example, in a more real implementation, the testing functionality would need a little bit more bells and whistles; if you wanted to test a function that does multiple mallocs, you'd want to be able to say "fail after X allocs on this thread", so that you could test all of them. Or you might want to do some sort of broad-scale integration test, where each allocation has a 1 in N chance of failure, determined by some customizable seed. Or other failures policies that are application-specific.

These seems like a lot of functionality to move into the malloc implementation itself (and would impose a maintenance burden elsewhere; we'd have to either add a config option, or do some careful dances to make sure we don't branch on testing_malloc_fail down fast paths).

I think, if anything, instead of supporting this directly we'd want to do it by allowing customizable hooks around the jemalloc entry points (which could be used to add fault injection, but could also do other useful things). The pain:reward ratio of putting the fault injection decisions in our repo instead of in user repos feels pretty high to me.

**Comment by jeamland** [comment]
Hrm. That makes sense. I'll have a go at spiking out an implementation of that.

Would {je_,}mallctl be the right place to set the hooks?

**Comment by davidtgoldblatt** [comment]
Yeah, that'd be the place. That said though, I'm still sort of skeptical of even that: We definitely don't want to take a branch on hook presence, and folding it behind fast path checks requires some tricky threading stuff. Getting the API right also feels hard without a couple different use cases.

Having a separate library (interposing the allocator API) that handles all the fault injection feels like it gets 90% of the benefits for 10% of the work.
